drivers/sensor/hx711/hx711.c

```diff
@@ -0,0 +1,562 @@
+ *
+ * @retval 0 on success
+ * @retval  -ENOTSUP if an invalid channel is given
+ *
+ */
+static int hx711_channel_get(const struct device *dev,
+			     enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct hx711_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_WEIGHT:
+	case SENSOR_CHAN_ALL:
+		val->val1 = sensor_value_to_double(&data->slope)
+			    * (data->reading - data->offset);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+/**
+ * @brief Initialise HX711.
+ *
+ * @param dev Pointer to the hx711 device structure
+ *
+ * @retval 0 on success
+ * @retval -EINVAL if an invalid argument is given
+ *
+ */
+static int hx711_init(const struct device *dev)
+{
+	LOG_DBG("Initialising HX711\n");
+
+	int ret = 0;
+	struct hx711_data *data = dev->data;
+	const struct hx711_config *cfg = dev->config;
+
+	LOG_DBG("SCK GPIO port : %s\n", cfg->sck_ctrl);
+	LOG_DBG("SCK Pin : %d\n", cfg->sck_pin);
+	LOG_DBG("DOUT GPIO port : %s\n", cfg->dout_ctrl);
+	LOG_DBG("DOUT Pin : %d\n", cfg->dout_pin);
+	LOG_DBG("RATE GPIO port : %s\n", cfg->rate_ctrl);
+	LOG_DBG("RATE Pin : %d\n", cfg->rate_pin);
+	LOG_DBG("Gain : %d\n", data->gain);
+	LOG_DBG("Offset : %d\n", data->offset);
+	LOG_DBG("Slope : %d.%d\n", data->slope.val1, data->slope.val2);
+
+	/* Configure SCK as output, LOW */
+	data->sck_gpio = device_get_binding(cfg->sck_ctrl);
+	if (data->sck_gpio == NULL) {
+		LOG_ERR("Failed to get GPIO device %s.", cfg->sck_ctrl);
+		return -EINVAL;
+	}
+	LOG_DBG("SCK pin controller is %p, name is %s\n",
+		data->sck_gpio,
+		data->sck_gpio->name);
+
+	ret = gpio_pin_configure(data->sck_gpio,
+				 cfg->sck_pin,
+				 GPIO_OUTPUT_INACTIVE | cfg->sck_flags);
+	if (ret != 0) {
+		return ret;
+	}
+
+#if DT_INST_NODE_HAS_PROP(0, rate_gpios)
+	/* Configure RATE as output, LOW */
+	data->rate_gpio = device_get_binding(cfg->rate_ctrl);
+	if (data->rate_gpio == NULL) {
+		LOG_ERR("Failed to get GPIO device %s.", cfg->rate_ctrl);
+		return -EINVAL;
+	}
+	LOG_DBG("RATE pin controller is %p, name is %s\n",
+		data->rate_gpio,
+		data->rate_gpio->name);
+	ret = gpio_pin_configure(data->rate_gpio,
+				 cfg->rate_pin,
+				 GPIO_OUTPUT_INACTIVE | cfg->rate_flags);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = gpio_pin_set(data->rate_gpio,
+			   hx711_config.rate_pin,
+			   CONFIG_HX711_SAMPLING_RATE);
+	if (ret != 0) {
+		return ret;
+	}
+#endif
+
+	/* Configure DOUT as input */
+	data->dout_gpio = device_get_binding(cfg->dout_ctrl);
+	if (data->dout_gpio == NULL) {
+		LOG_ERR("Failed to get GPIO device %s.", cfg->dout_ctrl);
+		return -EINVAL;
+	}
+	LOG_DBG("DOUT pin controller is %p, name is %s\n",
+		data->sck_gpio,
+		data->sck_gpio->name);
+	ret = gpio_pin_configure(data->sck_gpio,
+				 cfg->dout_pin,
+				 GPIO_INPUT | cfg->dout_flags);
+	if (ret != 0) {
+		return ret;
+	}
+	LOG_DBG("Set DOUT pin : %d\n", cfg->dout_pin);
+
+	/* Get a reading to set GAIN */
+	ret = hx711_sample_fetch(dev, SENSOR_CHAN_WEIGHT);
```

Comment:
Referencing https://github.com/zephyrproject-rtos/zephyr/pull/24295#issuecomment-779216345: This is problematic.

The init function is invoked by the kernel in a loop along with all other drivers at the same init level (`POST_KERNEL` in this case.)  Having this function stall for up to 60 s is not acceptable, as it would prevent startup from initializing other drivers and getting to the application.

Reply:
What is the progress on this right now?

Reply:
Unfortunately, after a powerdown or power reset the HX711 needs approximately 600ms to provide the first measurement (dout from high to low).

I could avoid calling **hx711_sample_fetch** during init but then there would not be any guarantee that the hardware is functioning properly.

Also I still would like to know what to do in case of bad wiring.
As is said as it is right now I get kernel panics during init if the module is not connected.

Not sure if this happens because 600ms is too much or because **-EACCESS** is returned from init.

Should we allow init to complete and get **-EACCESS** during the first call to **hx711_sample_fetch**?

Reply:
Never got an answer to this one. With the help of @nobodyguy we now use semaphores and interrupts, but the 600ms will still be there as it is hardware specific to HX711 as explained.

hx711_sample_fetch during init should not be avoided as it sets the appropriate GAIN and RATE according to KConfig.

I still get kernel panics if the sensor is not connected (bad wiring). This happens if **hx711_sample_fetch** returns an error (-EIO). 

I added an appropriate remark "Is the device properly connected?" but do not know how to circumvent this issue.

```
[00:00:01.034,332] <err> HX711: Weight data not ready within 700 ms. Is the device properly connected?
*** Booting Zephyr OS build v3.0.0-rc1-307-ga775606ae955  ***
[00:00:01.034,423] <inf> main: device is (nil), name is  /
[00:00:01.034,423] <inf> main: Calculating offset...
[00:00:01.034,454] <err> os: ***** MPU FAULT *****
[00:00:01.034,454] <err> os:   Data Access Violation
[00:00:01.034,454] <err> os:   MMFAR Address: 0x34a5
[00:00:01.034,484] <err> os: r0/a1:  0x04e000e1  r1/a2:  0x00000039  r2/a3:  0x00000003
[00:00:01.034,484] <err> os: r3/a4:  0x04e000e1 r12/ip:  0x00000000 r14/lr:  0x0000735d
[00:00:01.034,515] <err> os:  xpsr:  0x61000000
[00:00:01.034,515] <err> os: Faulting instruction address (r15/pc): 0x0000734c
[00:00:01.034,545] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:01.034,576] <err> os: Current thread: 0x200012b8 (main)
[00:00:09.903,228] <err> os: Halting system

```

Are errors during init not handled by the kernel?
Shouldn't they?
How can a driver return an error during init without causing kernel panics?